pkg/repository/prisma/workflow_run.go

```diff
@@ -838,6 +839,221 @@ func (w *workflowRunEngineRepository) GetScheduledChildWorkflowRun(ctx context.C
 
 	return w.queries.GetScheduledChildWorkflowRun(ctx, w.pool, childParams)
 }
+
+func (w *workflowRunEngineRepository) PopWorkflowRunsCancelInProgress(ctx context.Context, tenantId, workflowId string, maxRuns int) ([]*dbsqlc.WorkflowRun, []*dbsqlc.WorkflowRun, error) {
+	ctx, span := telemetry.NewSpan(ctx, "queue-by-cancel-in-progress")
+	defer span.End()
+
+	tx, commit, rollback, err := sqlchelpers.PrepareTx(ctx, w.pool, w.l, 15000)
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	defer rollback()
+
+	// place a FOR UPDATE lock on queued and running workflow runs to prevent concurrent updates
```

Comment:
ok just talking through this, so behavior here if a run finishes while this lock is acquired will force this cancelling state to be written but then the lock for success will be released putting it into a success state and the cancelled state will then be ignored?

Reply:
Discussed offline